packages/test/test-end-to-end-tests/src/test/localTestSignals.spec.ts

```diff
@@ -7,29 +7,89 @@ import { strict as assert } from "assert";
 
 import { describeCompat } from "@fluid-private/test-version-utils";
 import { ConnectionState } from "@fluidframework/container-loader";
-import { IInboundSignalMessage } from "@fluidframework/runtime-definitions/internal";
+import { IFluidDataStoreRuntime } from "@fluidframework/datastore-definitions/internal";
+import type {
+	IContainerRuntimeBase,
+	IInboundSignalMessage,
+} from "@fluidframework/runtime-definitions/internal";
 import {
 	DataObjectFactoryType,
 	ITestContainerConfig,
 	ITestFluidObject,
 	ITestObjectProvider,
 	getContainerEntryPointBackCompat,
-	timeoutPromise,
+	timeoutPromise as timeoutPromiseUnnamed,
+	TimeoutWithError,
+	TimeoutWithValue,
 } from "@fluidframework/test-utils/internal";
 
+type IContainerRuntimeBaseWithClientId = IContainerRuntimeBase & {
+	clientId?: string | undefined;
+};
+
+enum RuntimeLayer {
+	dataStoreRuntime = "dataStoreRuntime",
+	containerRuntime = "containerRuntime",
+}
+
+interface SignalClient {
+	dataStoreRuntime: IFluidDataStoreRuntime;
+	containerRuntime: IContainerRuntimeBaseWithClientId;
+	signalReceivedCount: number;
+	clientId: string | undefined;
+}
+
 const testContainerConfig: ITestContainerConfig = {
 	fluidDataObjectType: DataObjectFactoryType.Test,
 };
 
-const waitForSignal = async (...signallers: { once(e: "signal", l: () => void): void }[]) =>
-	Promise.all(
+async function timeoutPromise<T = void>(
+	executor: (controller: {
+		resolve: (value: T | PromiseLike<T>) => void;
+		reject: (reason?: any) => void;
+	}) => void,
+	timeoutOptions: TimeoutWithError | TimeoutWithValue<T> = {},
+): Promise<T> {
+	return timeoutPromiseUnnamed(
+		(resolve, reject) => executor({ resolve, reject }),
+		timeoutOptions,
+	);
+}
+
+async function waitForSignal(...signallers: { once(e: "signal", l: () => void): void }[]) {
+	return Promise.all(
 		signallers.map(async (signaller, index) =>
-			timeoutPromise((resolve) => signaller.once("signal", () => resolve()), {
+			timeoutPromise(({ resolve }) => signaller.once("signal", () => resolve()), {
 				durationMs: 2000,
 				errorMsg: `Signaller[${index}] Timeout`,
 			}),
 		),
 	);
+}
+
+async function waitForTargetedSignal(
+	targetedSignaller: { once(e: "signal", l: () => void): void },
+	otherSignallers: { once(e: "signal", l: () => void): void }[],
+) {
+	return Promise.all([
+		timeoutPromise(({ resolve }) => targetedSignaller.once("signal", () => resolve()), {
+			durationMs: 2000,
+			errorMsg: `Targeted Signaller Timeout`,
+		}),
+		otherSignallers.map(async (signaller, index) =>
+			timeoutPromise(
+				({ reject }) =>
+					signaller.once("signal", () =>
+						reject(`Signaller[${index}] should not have recieved a signal`),
+					),
+				{
+					durationMs: 100,
+					value: "No Signal Received",
+				},
+			),
```

Comment:
There are two ways to get errors from this logic. There is no code that seems to care which of the two happens, which are polar opposites in terms of behavior. I would say then that tests must rely on success, but there is no non-error path here. So, I feel like the tests aren't really testing what is desired.

Reply:
"No Signal Received" TimeoutWithValue is the desired non-error path here: https://github.com/microsoft/FluidFramework/blob/3d83ec9e7d07fef1b7cdb4cc5863ae9998f9642e/packages/test/test-utils/src/timeoutUtils.ts#L163C2-L174C16

```typescript
/**
 * @param timeoutOptions - Options that can be used to override the timeout and / or define the behavior when
 * the promise is not fulfilled. For example, instead of rejecting the promise, resolve with a specific value.
*/
```